src/components/VmConsole/VmConsoleSelector.js

```diff
@@ -0,0 +1,97 @@
+import React from 'react'
+import { connect } from 'react-redux'
+import PropTypes from 'prop-types'
+
+import { push } from 'connected-react-router'
+import { DropdownButton, MenuItem, Icon } from 'patternfly-react'
+import ConsoleConfirmationModal from '../VmActions/ConsoleConfirmationModal'
+import { MenuItemAction } from '../VmActions/Action'
+import { DOWNLOAD_CONSOLE } from '_/constants'
+import { msg } from '_/intl'
+import { getRDP } from '_/actions'
+
+import { isWindows } from '_/helpers'
+import style from './style.css'
+
+class VmConsoleSelector extends React.Component {
+  render () {
+    const { vmId, vms, consoles, config, consoleId, isConsolePage, onRDP } = this.props
+    let actions = vms.getIn(['vms', vmId, 'consoles'])
+    if (actions.size === 0) {
+      return <div />
+    }
+    const activeConsole = consoleId === 'rdp'
+      ? msg.rdpConsole()
+      : consoles.getIn(['vms', vmId, 'consoleStatus']) === DOWNLOAD_CONSOLE
+        ? msg[actions.find((a) => a.get('id') === consoleId).get('protocol') + 'Console']()
+        : msg.vncConsoleBrowser()
+    const vnc = actions.find((a) => a.get('protocol') === 'vnc')
+    const consoleItems = actions.map(action =>
+      <MenuItemAction
+        id={action.get('id')}
+        key={action.get('id')}
+        confirmation={<ConsoleConfirmationModal isConsolePage={isConsolePage} vm={vms.getIn(['vms', vmId])} consoleId={action.get('id')} onClose={() => {}} />}
+        shortTitle={msg[action.get('protocol') + 'Console']()}
+        icon={<Icon name='external-link' />}
+      />
+    ).toJS()
+
+    const hasRdp = isWindows(vms.getIn(['vms', vmId, 'os', 'type']))
+
+    if (hasRdp) {
+      const domain = config.get('domain')
+      const username = config.getIn([ 'user', 'name' ])
+      consoleItems.push(<MenuItem
+        key='rdp'
+        onClick={(e) => { e.preventDefault(); onRDP({ domain, username }) }}
+      >
+        {msg.rdpConsole()} <Icon name='external-link' />
+      </MenuItem>)
+    }
+
+    if (vnc.size) {
+      consoleItems.push(
+        <MenuItemAction
+          id={`${vnc.get('id')}_browser`}
+          key={`${vnc.get('id')}_browser`}
+          confirmation={<ConsoleConfirmationModal isConsolePage={isConsolePage} isNoVNC vm={vms.getIn(['vms', vmId])} consoleId={vnc.get('id')} onClose={() => {}} />}
+          shortTitle={msg.vncConsoleBrowser()}
+        />
+      )
+    }
+    return <div className={style['console-dropdown-box']}>
```

Comment:
Only suggestion for this component is to define the constants close to where they are used and use whitespace to show a "block level" distinction between the `consoleItems` pushes.